File: 2023/day12/php/solution.php
<?php
/**
 * Advent of Code 2023 Day 12: Hot Springs
 *
 * Uses memoized DP to count valid arrangements of operational and damaged springs.
 */

/**
 * Count valid arrangements using memoized dynamic programming.
 * State: (position, group_index, current_run_length)
 */
function countArrangements(string $pattern, array $groups): int {
    $memo = [];
    $patternLen = strlen($pattern);
    $groupsLen = count($groups);

    $dp = function(int $pos, int $groupIdx, int $currentRun) use (&$dp, &$memo, $pattern, $groups, $patternLen, $groupsLen): int {
        // Create memoization key
        $key = "$pos,$groupIdx,$currentRun";
        if (isset($memo[$key])) {
            return $memo[$key];
        }

        // Base case: reached end of pattern
        if ($pos === $patternLen) {
            // Valid if we've matched all groups and no partial run
            if ($groupIdx === $groupsLen && $currentRun === 0) {
                return 1;
            }
            // Or if we're on the last group and the run matches
            if ($groupIdx === $groupsLen - 1 && $groups[$groupIdx] === $currentRun) {
                return 1;
            }
            return 0;
        }

        $result = 0;
        $char = $pattern[$pos];

        // Option 1: Place operational spring (.)
        if ($char === '.' || $char === '?') {
            if ($currentRun === 0) {
                // No active run, just move forward
                $result += $dp($pos + 1, $groupIdx, 0);
            } elseif ($groupIdx < $groupsLen && $groups[$groupIdx] === $currentRun) {
                // End current run if it matches expected group size
                $result += $dp($pos + 1, $groupIdx + 1, 0);
            }
            // Otherwise invalid (run doesn't match group)
        }

        // Option 2: Place damaged spring (#)
        if ($char === '#' || $char === '?') {
            if ($groupIdx < $groupsLen && $currentRun < $groups[$groupIdx]) {
                // Can extend current run
                $result += $dp($pos + 1, $groupIdx, $currentRun + 1);
            }
            // Otherwise invalid (exceeds group size or no more groups)
        }

        $memo[$key] = $result;
        return $result;
    };

    return $dp(0, 0, 0);
}

/**
 * Parse a line into pattern and groups array.
 */
function parseLine(string $line): array {
    $parts = preg_split('/\s+/', trim($line));
    $pattern = $parts[0];
    $groups = array_map('intval', explode(',', $parts[1]));
    return [$pattern, $groups];
}

/**
 * Part 1: Sum of arrangement counts for all rows.
 */
function part1(array $lines): int {
    $total = 0;
    foreach ($lines as $line) {
        $line = trim($line);
        if (empty($line)) {
            continue;
        }
        [$pattern, $groups] = parseLine($line);
        $total += countArrangements($pattern, $groups);
    }
    return $total;
}

/**
 * Unfold pattern and groups by repeating them 5 times.
 */
function unfold(string $pattern, array $groups): array {
    $unfoldedPattern = implode('?', array_fill(0, 5, $pattern));
    $unfoldedGroups = [];
    for ($i = 0; $i < 5; $i++) {
        $unfoldedGroups = array_merge($unfoldedGroups, $groups);
    }
    return [$unfoldedPattern, $unfoldedGroups];
}

/**
 * Part 2: Sum of arrangement counts for all rows after unfolding.
 */
function part2(array $lines): int {
    $total = 0;
    foreach ($lines as $line) {
        $line = trim($line);
        if (empty($line)) {
            continue;
        }
        [$pattern, $groups] = parseLine($line);
        [$unfoldedPattern, $unfoldedGroups] = unfold($pattern, $groups);
        $total += countArrangements($unfoldedPattern, $unfoldedGroups);
    }
    return $total;
}

// Main execution
$input = file_get_contents(__DIR__ . '/../input.txt');
$lines = explode("\n", $input);

echo "Part 1: " . part1($lines) . "\n";
echo "Part 2: " . part2($lines) . "\n";

File: 2023/day12/rust/solution.rs
// Advent of Code 2023 Day 12: Hot Springs
// Rust solution using memoized dynamic programming

use std::collections::HashMap;
use std::fs;

fn count_arrangements(pattern: &[u8], groups: &[usize]) -> i64 {
    let mut memo: HashMap<(usize, usize, usize), i64> = HashMap::new();
    dp(pattern, groups, 0, 0, 0, &mut memo)
}

fn dp(
    pattern: &[u8],
    groups: &[usize],
    pos: usize,
    group_idx: usize,
    current_run: usize,
    memo: &mut HashMap<(usize, usize, usize), i64>,
) -> i64 {
    // Check memoization
    let key = (pos, group_idx, current_run);
    if let Some(&result) = memo.get(&key) {
        return result;
    }

    // Base case: reached end of pattern
    if pos == pattern.len() {
        let result = if group_idx == groups.len() && current_run == 0 {
            // Valid: matched all groups and no partial run
            1
        } else if group_idx == groups.len() - 1 && groups[group_idx] == current_run {
            // Valid: on last group and run matches
            1
        } else {
            0
        };
        memo.insert(key, result);
        return result;
    }

    let mut result = 0;
    let char = pattern[pos];

    // Option 1: Place operational spring (.)
    if char == b'.' || char == b'?' {
        if current_run == 0 {
            // No active run, just move forward
            result += dp(pattern, groups, pos + 1, group_idx, 0, memo);
        } else if group_idx < groups.len() && groups[group_idx] == current_run {
            // End current run if it matches expected group size
            result += dp(pattern, groups, pos + 1, group_idx + 1, 0, memo);
        }
        // Otherwise invalid (run doesn't match group)
    }

    // Option 2: Place damaged spring (#)
    if char == b'#' || char == b'?' {
        if group_idx < groups.len() && current_run < groups[group_idx] {
            // Can extend current run
            result += dp(pattern, groups, pos + 1, group_idx, current_run + 1, memo);
        }
        // Otherwise invalid (exceeds group size or no more groups)
    }

    memo.insert(key, result);
    result
}

fn parse_line(line: &str) -> (Vec<u8>, Vec<usize>) {
    let parts: Vec<&str> = line.trim().split_whitespace().collect();
    let pattern = parts[0].as_bytes().to_vec();
    let groups: Vec<usize> = parts[1].split(',').map(|x| x.parse().unwrap()).collect();
    (pattern, groups)
}

fn unfold(pattern: &[u8], groups: &[usize]) -> (Vec<u8>, Vec<usize>) {
    let mut unfolded_pattern = Vec::new();
    for i in 0..5 {
        if i > 0 {
            unfolded_pattern.push(b'?');
        }
        unfolded_pattern.extend_from_slice(pattern);
    }
    let unfolded_groups: Vec<usize> = groups.iter().cycle().take(groups.len() * 5).cloned().collect();
    (unfolded_pattern, unfolded_groups)
}

fn part1(lines: &[&str]) -> i64 {
    let mut total = 0;
    for line in lines {
        if line.trim().is_empty() {
            continue;
        }
        let (pattern, groups) = parse_line(line);
        total += count_arrangements(&pattern, &groups);
    }
    total
}

fn part2(lines: &[&str]) -> i64 {
    let mut total = 0;
    for line in lines {
        if line.trim().is_empty() {
            continue;
        }
        let (pattern, groups) = parse_line(line);
        let (unfolded_pattern, unfolded_groups) = unfold(&pattern, &groups);
        total += count_arrangements(&unfolded_pattern, &unfolded_groups);
    }
    total
}

fn main() {
    let input = fs::read_to_string("../input.txt").expect("Failed to read input file");
    let lines: Vec<&str> = input.lines().collect();

    println!("Part 1: {}", part1(&lines));
    println!("Part 2: {}", part2(&lines));
}
